### PR TITLE
Add adjoint state gradients tests

### PR DESCRIPTION
**Pull request by gautierronan**
- https://github.com/pierreguilmin/torchqdynamics/pull/77/commits/042fa551272ab0643071ad6f5bc08af2b4e7d748 modifies the `LeakyCavity` model to allow for parameters
- https://github.com/pierreguilmin/torchqdynamics/pull/77/commits/9ed3da426b652cebefe8956f090c6c61275111ca implements the adjoint tests and fixes the tests

### CODE REVIEW

File: tests/mesolve/open_system.py [hunk 1]
```diff
@@ -34,31 +34,62 @@ class LeakyCavity(OpenSystem):
...
         alpha_t = (
-            self.alpha0 * np.exp(-1j * self.delta * t) * np.exp(-0.5 * self.kappa * t)
+            self.alpha0
+            * torch.exp(-1j * self.delta * t)
+            * torch.exp(-0.5 * self.kappa * t)
         )
         return tq.coherent_dm(self.n, alpha_t)
+
+    def loss(self, rho: Tensor) -> float:
+        """Compute the loss function of a given density matrix."""
```
Comment (pierreguilmin) [inline tests/mesolve/open_system.py:93]:
```suggestion
        """Compute an example loss function from a given density matrix."""
```

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
Very nice!!

File: tests/mesolve/mesolver_tester.py [hunk 2]
```diff
@@ -72,3 +72,46 @@ def _test_rho_save(
...
+
+class AdjointMESolverTester(MESolverTester):
+    def _test_adjoint(
+        self,
+        options: Options,
+        system: OpenSystem,
+        *,
+        num_t_save: int,
+        atol: float = 1e-3,
+        rtol: float = 1e-5,
```
Comment (pierreguilmin) [inline tests/mesolve/mesolver_tester.py:85]: Could we switch the order here and in the call to `torch.allclose`?

File: tests/mesolve/mesolver_tester.py [hunk 3]
```diff
@@ -72,3 +72,46 @@ def _test_rho_save(
 
     def test_rho_save(self):
         pass
+
+
+class AdjointMESolverTester(MESolverTester):
```
Comment (pierreguilmin) [inline tests/mesolve/mesolver_tester.py:77]:
```suggestion
class MEAdjointSolverTester(MESolverTester):
```
It's a _Master Equation Adjoint Solver_ and not an _Adjoint Master Equation Solver_.

**Review by pierreguilmin — APPROVED** [review]
Nitpicking.
